### PR TITLE
alchemy scaffold updates

### PR DESCRIPTION
**Pull request by mmerickel**
This is a WIP to bring the alchemy scaffold in line with best practices.

- [x] Remove scoped sessions.
- [x] Connect session to request method.
- [x] Switch to jinja2 templates.
- [x] Add a `meta` module or package to avoid any circular references when importing.
- [x] Switch models into a package instead of a module.
- [x] Scan docs and tutorials for updates related to the changes. Here is the list of files. [See comment below](https://github.com/Pylons/pyramid/pull/2024#issuecomment-150054378).
  - [x] whatsnew-1.6.rst (Scaffolding Enhancements and Documentation Enhancements sections)
  - [x] Quick Tour `sqla_demo` generated from scaffold and its associated files in `quick_tour/sqla_demo/`. Note that there are some comments inserted in these files for Sphinx rendering. See quick_tour.rst, Databases section.
  - [x] quick_tour.rst, Databases section
  - [x] tutorials/wiki2/src/\*/tutorial/[~~authorization~~|~~basiclayout~~|~~models~~|~~tests~~|~~views~~]/ (5 directories)
  - [x] tutorials/wiki2/authorization.rst (13 occurrences)
  - [x] tutorials/wiki2/basiclayout.rst (2 occurrences)
  - [x] tutorials/wiki2/definingmodels.rst (9 occurrences)
  - [x] tutorials/wiki2/definingviews.rst (9 occurrences)
  - [x] tutorials/wiki2/design.rst (4 occurrences)
  - [x] tutorials/wiki2/tests.rst
- Pyramid Cookbook recipes
  - [ ] http://docs.pylonsproject.org/projects/pyramid-cookbook/en/latest/auth/wiki2_auth.html (Add a note that this recipe will no longer apply when Pyramid 1.7 is released, and is most likely out of date since Pyramid 1.3. Or maybe just delete it?)
- [x] Add `configure_mappers` when initializing the session.

### CODE REVIEW

**Comment by stevepiercy** [comment]
I'll take on this todo, and list all the files:
[ ] Scan docs and tutorials for updates related to the changes.

**Comment by bertjwregeer** [comment]
Woah... switch to Jinja 2 templates?

**Comment by blaflamme** [comment]
I would have updated for mako. I don't think getting closer to flask/django is a good thing in the end. I would prefer getting closer to mike's stuff for our defaults if we have to make a choice.

**Comment by mmerickel** [comment]
I don't mind using jinja2. @pauleveritt uses it in the quick tutorial and I use it in all my projects. Since when is this "closer to flank/django" a good or bad thing? It's a templating language.

I mean if people feel so strongly about it that's fine but I told @ergo he could use what he wanted since he was willing to do the work.

**Comment by ergo** [comment]
@blaflamme I was having a hard time to decide on mako/jinja2 since I know and use both for a long time - the reason why I decided on jinja2 was to use something that has a syntax that people are more used to - this lowers the entry barrier a bit which I find a good thing because we want to have more people use pyramid in the end.
If there is a big opposition I don't mind changing to mako instead.

**Comment by pauleveritt** [comment]
If we’re going to vote for unique, less-popular templating languages, we should go with the one ChrisM prefers, Chameleon. It’s the original template language for Pyramid. But I don’t think you’d like that choice. [wink] 

At the end of the day, we all have different ideas, but Jinja2 and its syntax (Django) are far more popular. We shouldn’t make templating the place where newbies (the audience for a scaffold) thinks we are “different”. Wrong battle to pick, IMO.

—Paul

**Comment by blaflamme** [comment]
@ergo no need to change, I wondered why.

**Comment by blaflamme** [comment]
@pauleveritt sure, I don't want to bring any battle to the table ;)

But I would have naturally coupled the zodb scaffold with chameleon and the sqla scaffold with mako.

**Comment by piotr-dobrogost** [comment]
How about giving choice and making template system be a parameter of scaffolds? In the end scaffold should be opinionated in regard to its core functionality (SQLAlchemy in this case) and not auxiliary things (template system in this case).

**Comment by ergo** [comment]
@piotr-dobrogost This might be problematic from maintenance perspective since we would have to understand and update all the templates simulatenously whenever anything would change.

@blaflamme At first my initial idea was to use mako, but because of the reasons @mmerickel and @pauleveritt mentioned it felt like the best option to go for jinja2.

**Comment by stevepiercy** [comment]
The suggestion from @piotr-dobrogost would inform newbies that there are several templating languages. It would also mean more development work, including the hardest task of all, naming things, compounded by the backward compatible alias of `--template` to `--scaffold` in pcreate. /play noooo

**Comment by mmerickel** [comment]
Showing multiple templating systems is great but also overly complex for the purposes of this scaffold and PR. Some future work to create alternate start scaffolds would probably be interesting though. ``pcreate -s starter-mako`` etc.

**Comment by ergo** [comment]
http://cookiecutter.readthedocs.org/en/latest/ - probably has all the functionality needed.

I've been doing some workshops recently introducing pyramid for people, and after got some feedback - especially when evangelizing to people who only did django before - jinja2 seems like a good choice.

**Comment by pauleveritt** [comment]
I reached the same conclusion a while back. @cguardia is doing a lot of work on this in https://github.com/cguardia/cookiecutter-pyramid

—Paul

**Comment by blaflamme** [comment]
Same for me, I've been using cookiecutter for a while for my own stuff and prefer it over scaffolds. We could eventually have official cookiecutters maintained separately from the pyramid core, like we did with template stuff.

**Comment by bertjwregeer** [comment]
Questions have come up if we want to add alembic migrations to the scaffold. I'd vote for yes, but I am biased in that I have 4 different projects, and I have implemented migrations 4 different ways, having one way that is standardised would be nice.

**Comment by stevepiercy** [comment]
What am I doing wrong in my git workflow? I want the commits in yonder PR #2103 to show up in this here PR #2014.

Here's what I did:
1. Checked out the branch Pylons/pyramid/feature/alchemy-scaffold-update
2. Made changes, committed, pushed to my fork of the repo
3. Compared: https://github.com/Pylons/pyramid/compare/feature/alchemy-scaffold-update...stevepiercy:feature/alchemy-scaffold-update
4. Made a PR off the comparison

Instead of step 2, should I have pushed to Pylons/pyramid/feature/alchemy-scaffold-update?

**Comment by bertjwregeer** [comment]
@stevepiercy You need to click the merge pull request button on #2103 and it will show up here. Just did that. Looking forward to seeing these changes :-)

**Comment by stevepiercy** [comment]
I don't know git-fu to resolve the conflicts on GitHub, and I don't trust myself to make it worse by following instructions under the link for "Use the command line". I need a little help.

**Comment by tseaver** [comment]
@stevepiercy I merged `master` into `feature/alchemy-scaffold-update` and pushed it.  Once Travis passes, it should be possible to merge from the Github UI.

**Comment by mmerickel** [comment]
@tseaver Your merge is not correct. The entire branch is gone now because of the common ancestors created by @stevepiercy merging it into master and then reverting it.

**Comment by mmerickel** [comment]
This was never supposed to be merged into master, it's in a pretty bad state right now.

**Comment by bertjwregeer** [comment]
On this branch, we need a revert for the revert to bring back all the commits that were inadvertently pushed to `master`.

It's a mess because stuff was accidentally pushed to `master`, not much we can do about it though :/

**Comment by stevepiercy** [comment]
I was uncertain of the content I added to `tutorials/wiki2/basiclayout.rst`. I'd appreciate a second pair of eyes on that one from someone who has a better command of translating Python to narrative English.

**Comment by pauleveritt** [comment]
You are really kicking butt on all this, Steve. Kudos.

Some comments:

- "in which it's contained” -> its

- You have two paragraphs that start with “Next” (following a paragraph that says “The next step”)

- This is a good way to say it: "@view_config is inert without being detected via a :term:`scan`.”

- "The models package is where the alchemy scaffold put the classes” -> “puts”?

- Perhaps reverse these two sentences:

“””
The Index import and the Index object creation is not required for this tutorial, and will be removed in the next step.

That's about all there is to it regarding models, views, and initialization code in our stock application.
“”"

**Comment by davisagli** [comment]
> - "in which it's contained” -> its

No. It's = "it is" which is correct here.

**Comment by pauleveritt** [comment]
Crap, you’re right…I thought possessive, but it’s the contraction.

—Paul

**Comment by stevepiercy** [comment]
Thanks @pauleveritt. I made a few changes. I went with the past tense "put", since we created the project from the scaffold. I was more concerned about my Pyton-to-English translation of the new `views` and `models` packages. :grinning:

**Comment by stevepiercy** [comment]
I started work on this step of the wiki tutorial:
http://docs.pylonsproject.org/projects/pyramid/en/latest/tutorials/wiki2/authorization.html

Should I keep security as a single file `security.py` or make a subpackage `security`?

While I'm at it, should I do something similar with `tests.py`?

I don't have an opinion either way, so I'll stick with the existing examples unless someone can offer a reason to change it.

**Comment by bertjwregeer** [comment]
One of the first things I do after I create a project with the scaffold (and really I should just update my personal scaffolds ;-)) is to move almost everything to sub packages.

**Comment by stevepiercy** [comment]
@ergo or anyone else familiar with tests on this new branch: I could use help rewriting the tests for the SQLAlchemy wiki tutorial for this branch, and out of that process, better document what is going on within the tests. [The current documentation of the tests assumes too much knowledge for newbs](http://docs.pylonsproject.org/projects/pyramid/en/latest/tutorials/wiki2/tests.html). Most of the tests in the current documentation are self-explanatory by their names, but I am lost as to how to rewrite `_initTestingDB` for this new branch. Or should I reuse the new `BaseTest` class in this new branch, and if so, how?

Thanks in advance for holding my hand through this.

**Comment by ppaez** [comment]
@stevepiercy there is some progress in rewriting the tests in #2144; the functional tests are not passing, I haven't figured out yet how to initialize the application's in-memory database.  I tried using BaseTest, without success.

**Comment by ztane** [comment]
Let's just all switch to [Tonnikala](https://github.com/tetframework/Tonnikala) :innocent:

File: docs/quick_tour/sqla_demo/sqla_demo/models/meta.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from sqlalchemy import engine_from_config
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.schema import MetaData
+import zope.sqlalchemy
+
+NAMING_CONVENTION = {
```
Comment (miohtama) [inline docs/quick_tour/sqla_demo/sqla_demo/models/meta.py]: Maybe a comment here why this name convention has been chosen as it is unclear from the context and novice like me may ponder it. Is it arbitrary choice by the authors, why not SQLAlchemy default or does it base itself on some other existing convention?

**Comment by ztane** [comment]
Indeed, I've seen that naming convention in SQLAlchemy-Continuum which is written by a Flask Fanboy so I suspect this is from Flask? In any case it is just extra clutter for 99 % of people.

File: docs/quick_tour/sqla_demo/sqla_demo/models/meta.py (hunk 1 above)
Comment (ztane) [inline docs/quick_tour/sqla_demo/sqla_demo/models/meta.py]: Guess this is from Flask. Unnecessary clutter to 99 % of people I'd say.

Comment (pauleveritt) [inline docs/quick_tour/sqla_demo/sqla_demo/models/meta.py]: > On Dec 1, 2015, at 3:21 AM, Antti Haapala <notifications@github.com> wrote:
> 
> In docs/quick_tour/sqla_demo/sqla_demo/models/meta.py <https://github.com/Pylons/pyramid/pull/2024#discussion_r46250265>:
> 
> > @@ -0,0 +1,46 @@
> > +from sqlalchemy import engine_from_config
> > +from sqlalchemy.ext.declarative import declarative_base
> > +from sqlalchemy.orm import sessionmaker
> > +from sqlalchemy.schema import MetaData
> > +import zope.sqlalchemy
> > +
> > +NAMING_CONVENTION = {
> Guess this is from Flask. Unnecessary clutter to 99 % of people I'd say.
> 
Actually it’s from Alembic.

—Paul

**Comment by ergo** [comment]
@ztane No this is not extra clutter - there is a very good explanation for it to be present and it should not be removed.

http://alembic.readthedocs.org/en/latest/naming.html

On some databases you will end up with constraint names like "SYS_C0029334".  This is currently suggested naming convention by sqlalchemy and alembic.

File: docs/quick_tour/sqla_demo/sqla_demo/models/meta.py (hunk 1 above)
Comment (bertjwregeer) [inline docs/quick_tour/sqla_demo/sqla_demo/models/meta.py]: Drop a comment with a link to http://alembic.readthedocs.org/en/latest/naming.html right above it.

**Comment by ztane** [comment]
I stand corrected. Then its usage should be commented in there (in any case, it is easier to delete the unnecessary comments than to find such information if needed).

**Comment by enomado** [comment]
Just interested. Why are you removed sqlalchemy's `scoped sessions`  ?

**Comment by mmerickel** [comment]
@enomado You may bring them back in your own code if you like but we want to emphasize that they are unnecessary and better practices exist with regards to control over the lifecycle of your sessions.

**Comment by bertjwregeer** [comment]
Did some gardening, merged master into this PR and resolved all conflicts.

**Comment by bertjwregeer** [comment]
Could we look at pulling https://github.com/Pylons/pyramid/pull/1625 into this as well please?

**Comment by mmerickel** [comment]
Hey everyone, I'm making a few minor changes to the scaffold codebase and attempting to update the tutorials to track those changes in #2334. Just wanted to let people know incase they wanted to provide input. My plan is to finish that up and merge everything together into the final version for the 1.7 release!

**Comment by mmerickel** [comment]
http://docs.pylonsproject.org/projects/pyramid/en/feature-alchemy-scaffold-update/tutorials/wiki2/index.html

I've merged in #2334 which includes major updates to the scaffold and the tutorial and also enabled RTD to build this branch to make it easy to review.

@pauleveritt and @ergo and everyone else please take a look when you can so we can get this into master and ready for 1.7!

**Comment by ztane** [comment]
dbsession gets my vote as well, or sqla_session; after all it is the *session* as opposed to *engine*, *connection* or myriad other sqla concepts.

**Comment by piotr-dobrogost** [comment]
I agree with @ztane. Simple `request.sql` goes directly against *Explict is better than implicit* aphorism. I think `get_sql_session_factory` and `request.sql_session` would be explicit and coherent.

**Comment by ergo** [comment]
I would go with `request.sa_session`

**Comment by stevepiercy** [comment]
* The thing we get and use is a _session_. I'm with @ztane on this part. I'm :+1: for `session` in the name.
* _SA_ provides the session to the database. I'm :+1: for `sa` in the name.
* The _db_ does not provide directly the session. I'm :-1: for `db` in the name.
* SQL is a _language_. I'm :-1: for `sql` in the name.

`request.sa_session` resonates most with me. When reading aloud, it sounds like two words "ESS-ay SESS-shun" and reinforces what it is, namely a SQLAlchemy session.

**Comment by mmerickel** [comment]
Hey folks, I put my feelings on the `request.dbsession` rename in https://github.com/Pylons/pyramid/pull/2024#commitcomment-16223882.

Major omissions from this round of tutorial and scaffold changes:

- No alembic configuration.
- No py.test tests.
- The tests themselves are fairly weakly organized but they do work and so far there hasn't been a lot of interest in improving them.

These omissions are things I'm okay with and would love to get it out there in 1.7.

I think this is about ready to merge so any reviews on what's there other than just the name issue would be great!

**Comment by stevepiercy** [comment]
Would it be too much at this time to update the docs to use pip instead of setuptools on this branch, or should we wait until this branch is merged to master?

**Comment by bertjwregeer** [comment]
Update the docs @stevepiercy  :-)

**Comment by stevepiercy** [comment]
I found that until this scaffold is released, I cannot use it as a regular consumer would. Instead I must follow the HACKING.txt procedure of `git clone; setup.py dev`. Thanks to @mcdonc for talking it through with me.

I am updating the docs using .rst comments for what I think would be correct for pip. Upon a release of 1.7, we could uncomment the commented steps, test the steps, then do another release at 1.7.1.

**Comment by mmerickel** [comment]
@stevepiercy please revert your pip changes. You may keep in the enhancements and grammar checks but this branch is not the appropriate place to start changing everything to pip.

**Comment by stevepiercy** [comment]
@mmerickel will do.

Given the [chicken-and-egg problem](https://github.com/Pylons/pyramid/pull/2024#issuecomment-186798104), what do you suggest for the pip changes? Using .rst comments was a compromise: they won't show on the rendered docs, but doc writers can see them.

**Comment by bertjwregeer** [comment]
Sorry, that's my bad :-(

**Comment by mmerickel** [comment]
We need to begin re-writing the narrative docs to use pip. It doesn't need to start with this tutorial. That work can be done in another PR and when this is merged we can then begin updating the tutorial to use pip as well. I suggest starting a new WIP branch in the pyramid repo for all of that work.

As it stands we could release 1.7 with this branch as there's no telling how long it will take to update the entire repo to pip. (If we can finish view derivations I'm likely to release a 1.7 with that and this branch and have that be it, leaving pip for 1.8 depending on how that work progresses!)

**Comment by mmerickel** [comment]
So far I haven't heard anyone +1 or -1 this PR which is most of the reason I haven't merged it to master yet. Some review of the content would be great so we can move on from this massive PR.

**Comment by stevepiercy** [comment]
@mmerickel I'm focusing on reviewing docs, which might take me a couple of days, before ±1-ing it. Is there an urgency to merge sooner than later?

**Comment by mmerickel** [comment]
@stevepiercy the only urgency is to move on from this giant PR. Once it has achieved its goals it begins to stagnate if we don't merge it - the life of any long-lived branch. Also things like your pip revamp are obviously dependent on it.

**Comment by pauleveritt** [comment]
I have looked through it but not in depth, sorry. I yield to your good taste. Let's merge it.

**Comment by mmerickel** [comment]
@pauleveritt No worries I'm at least going to wait for @stevepiercy to run through it. @ergo said in #pyramid that "... I think the branch gets +1 from me to merge, everything else looks good". His main concern was about the jinja2 templates using `../templates/foo.jinja2` but I don't think that needs to block this PR.

**Comment by bertjwregeer** [comment]
I took @stevepiercy's commits and removed all references to pip, but kept his changes for grammar/punctuation and other improvements.

**Comment by bertjwregeer** [comment]
Just went through, didn't do a full follow along, but this looks really good. :+1: from me.

**Comment by stevepiercy** [comment]
> I suggest starting a new WIP branch in the pyramid repo for all of that work.

The branch is [docs/easy-install-to-pip](https://github.com/Pylons/pyramid/tree/docs/easy-install-to-pip)

**Comment by stevepiercy** [comment]
@mmerickel when running the tests on py3.5, I got three deprecation warnings, although the tests passed:

```bash
$ $VENV/bin/python setup.py test -q
running test
running egg_info
writing top-level names to tutorial.egg-info/top_level.txt
writing requirements to tutorial.egg-info/requires.txt
writing entry points to tutorial.egg-info/entry_points.txt
writing dependency_links to tutorial.egg-info/dependency_links.txt
writing tutorial.egg-info/PKG-INFO
reading manifest file 'tutorial.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*.cfg'
warning: no files found matching '*.rst'
warning: no files found matching '*.ico' under directory 'tutorial'
warning: no files found matching '*.gif' under directory 'tutorial'
warning: no files found matching '*.jpg' under directory 'tutorial'
warning: no files found matching '*.pt' under directory 'tutorial'
warning: no files found matching '*.txt' under directory 'tutorial'
warning: no files found matching '*.mak' under directory 'tutorial'
warning: no files found matching '*.mako' under directory 'tutorial'
warning: no files found matching '*.js' under directory 'tutorial'
warning: no files found matching '*.html' under directory 'tutorial'
warning: no files found matching '*.xml' under directory 'tutorial'
writing manifest file 'tutorial.egg-info/SOURCES.txt'
running build_ext
/Users/stevepiercy/projects/alchemy-scaffold/env/lib/python3.5/site-packages/docutils-0.12-py3.5.egg/docutils/io.py:236: DeprecationWarning: 'U' mode is deprecated
  self.source = open(source_path, mode, **kwargs)
.................../Users/stevepiercy/projects/alchemy-scaffold/tutorial/tutorial/views/default.py:31: DeprecationWarning: cgi.escape is deprecated, use html.escape instead
  return '<a href="%s">%s</a>' % (add_url, cgi.escape(word))
/Users/stevepiercy/projects/alchemy-scaffold/tutorial/tutorial/views/default.py:28: DeprecationWarning: cgi.escape is deprecated, use html.escape instead
  return '<a href="%s">%s</a>' % (view_url, cgi.escape(word))
..
----------------------------------------------------------------------
Ran 21 tests in 5.117s

OK
```

Changing `cgi.escape` to `html.escape` in `views/default.py` resolved two of the three warnings. Should I put these two changes into a PR?

I have no clue how to resolve the docutils warning.

**Comment by mmerickel** [comment]
@stevepiercy I'm not worried about issues with deprecation warnings introduced in py35 right now. It's better that the tutorial works on py27 and I'd like to avoid hacks in the code to work on both without warnings since our templates don't allow conditional rendering right now. Short answer - don't worry about it.

**Comment by stevepiercy** [comment]
@mmerickel please review #2385 for the feature to redirect to edit page when user attempts to add page that already exists.

**Comment by ztane** [comment]
[`cgi.escape`](https://docs.python.org/3/library/cgi.html#cgi.escape) was deprecated in Python **3.2**, for the reason that it does not escape quotes by default. Since jinja2 is used there, the code could as well teach better habits and use for example `markupsafe.escape()`

**Comment by mmerickel** [comment]
@ztane Sure, my point was that I don't want the code forked on py2 vs py3. If there is a library available to do the escaping on both without warnings then fine. If you or anyone else wants to make that change then be sure to add markupsafe as a dependency in setup.py next to docutils and the text explaining why.

**Comment by mmerickel** [comment]
Awesome folks. Any further changes such as markupsafe should be made as PRs to master! Thanks for the work everyone put into this since Pycon 2014.

**Comment by pauleveritt** [comment]
Wow, merged to master. Mega kudos to @mmerickel and also @stevepiercy for keeping an eye on all the ramifications.

**Comment by bertjwregeer** [comment]
Thank you @mmerickel and @stevepiercy!

**Comment by reedobrien** [comment]
applause

**Pull request merged**